RFC: Sub Team Maintainer (formerly author) Creates Repo Issues (#106)
buildpacks/rfcs: text/0000-create-repo-issues.md

# Definitions
[definitions]: #definitions

* RFC &rarr; Request For Comments. It is the process followed by the Cloud Native Buildpacks project, as well as many other OSS ones, to provide a consistent and controlled path for new features to enter the spec, libraries, or tools, so that all stakeholders can be confident about the direction Cloud Native Buildpacks is evolving in.
* FCP &rarr; Final Comment Period. This is step taken on a particular RFC when enough discussion of the tradeoffs have taken place and the team is in a position to make a decision. Before entering FCP, super majority of the team must sign off.
* PR &rarr; (Github) Pull Request.

# Motivation
[motivation]: #motivation

As the number of RFCs increase, it has become increasingly hard to ensure that a given RFC has been implemented, or will be implemented. This RFC proposes an amendment to the [RFC process](https://github.com/buildpacks/rfcs#rfc-process).

The current RFC process [defines](https://github.com/buildpacks/rfcs#rfc-process) that:
> the team member who merges the pull request should do the following ... Create a corresponding issue in the appropriate repo.

> In practice, that adds extra work for the team member who merges in the PR, and it may slip through the cracks occasionally. Additionally, the author of the RFC, as the one who proposed the change, is often the best person to create the issue, as the person who is most invested in that change. 

Reviewer: I am not sure that I agree that "the person who proposed the change is the best person to create the issue". I fear that adding requirements like this makes the RFC process even more unapproachable for new contributors. I believe the best person to create the issue is a maintainer of the relevant repo. 

As an implementation maintainer I currently try to create spec and lifecycle issues as needed when RFCs merge (examples: https://github.com/buildpacks/spec/issues/106, https://github.com/buildpacks/lifecycle/issues/354).

This makes sense to me b/c I have the lifecycle knowledge and permissions required to subdivide, label, organize in milestones etc. Example: RFC-0053 affected two APIs therefore needs two separate spec issues ( https://github.com/buildpacks/spec/issues/134, https://github.com/buildpacks/spec/issues/133) I don't think think every RFC author should have to think about that level of detail (although stephen wrote this one and he certainly could have ;p).

On the other hand I _do_ think every subteam maintainer should be thinking about what each RFC means for their subteam as RFCs go through and should be planning/organizing accordingly.

Reviewer 2: This mirrors my comment above. I'm very much in favor of issues being created from RFCs, but I think this should be handled by the subteams (not just the maintainers). Would you be open to that change @dfreilich?

Author: I'd definitely be open to that, that makes sense. One suggestion - do people think it makes sense to add additional label(s), like `needs-issues-impl`, `needs-issues-platform`, etc? In that context, when an RFC is put in FCP, the appropriate labels could be added, and they'd need to be removed for an RFC to be merged in. Or would that be over-engineering the solution...

- [EmberJS RFC Process](https://github.com/emberjs/rfcs#the-rfc-life-cycle) &rarr; `Once an RFC becomes active the relevant teams will plan the feature and create issues in the relevant repositories.`

A number of other RFC processes don't define who creates the issues, though they do emphasize that while the author of an RFC is not obligated to implement it, a given RFC having been accepted implies nothing about what priority is assigned to its implementation, nor whether anybody is currently working on it.
Some of those include: 
* [GatsbyJS](https://github.com/gatsbyjs/gatsby/tree/master/rfcs#the-rfc-life-cycle)
* [NixOS](https://github.com/NixOS/rfcs#the-rfc-life-cycle)
* [Yarn](https://github.com/yarnpkg/rfcs#the-rfc-life-cycle)
* [React](https://github.com/reactjs/rfcs/blob/master/README.md#the-rfc-life-cycle)

[Swift](https://github.com/apple/swift-evolution/blob/master/process.md#preparing-an-implementation) requires that an RFC be coupled with a PR for its implementation.

# Unresolved Questions
[unresolved-questions]: #unresolved-questions

> # Spec. Changes (OPTIONAL)

Reviewer 2: This detail is great, but as a FYI the RFC process is not part of the spec so would not need a spec change.

The current RFC process [defines](https://github.com/buildpacks/rfcs#rfc-process) that:
> the team member who merges the pull request should do the following ... Create a corresponding issue in the appropriate repo.

In practice, that adds extra work for the team member who merges in the PR, and it may slip through the cracks occasionally. Additionally, the sub-team maintainers are often the best people to create the issue, given their insight into how it should be implemented.

As an additional consideration, it is often hard to track work from the RFC proposal to the implementation, and difficult to ensure that it was completed. If we use [Github references](https://docs.github.com/en/github/writing-on-github/autolinked-references-and-urls#issues-and-pull-requests) to link the RFC PR to the issue on the appropriate repository, we can better track the work, and be notified when it is fully implemented.

# What it is
[what-it-is]: #what-it-is
This defines a process for sub-team maintainers to open issues on the relevant repositories, after FCP for the RFC concludes, before it is merged in.

# How it Works
[how-it-works]: #how-it-works

> When a PR has been accepted (after FCP has concluded), the sub-team maintainers make issues (tagging the RFC PR) in the relevant buildpack repos, and labels the PR `issues-created`. At that point, the RFC PR may be merged in.

Reviewer: Given that some sub-teams have multiple maintainers, there may still be some potential for ambiguity around who will handle a particular RFC. I propose that when we move RFCs into FCP we should assign a sub-team maintainer to the PR and they will be responsible for creating the issues.

Reviewer 2: @ekcasey my thought is that this would happen during the weekly subteam meetings, but I don't feel strongly about it.

Author: @ekcasey I don't know whether we need that level of process at this point. I agree with @hone that I had felt it could be done during the subteam meetings, as one of the repeating action items. Should I clarify that?

Reviewer 3: There are some RFCs that require multiple sub-teams to create issues from a single RFC. For the sake of minimizing coordination could we maybe change this to be a label per subteam.

For example:

* `issues-created/platform`
* `issues-created/implementation`

If no issues need to be created for a given subteam they can add that label for brevity sake. We can further automate this process by checking that these labels are present.

Author: I incorporated @jromero's feedback. LMK what you think.

Given an RFC PR #002, which discusses a change relevant to the `pack CLI` to add in a `pack my-bags` command. After discussion in Working Group meetings and votes from the Core Team, the RFC is placed in FCP for 7 days. At the end of those, the platform maintainers should open an issue in the `pack CLI`, which tags the RFC PR. For example,

```
### Description

To enable easier traveling, we would like to add a `pack my-bags` command

### Proposed solution
This issue is a request to implement the proposed solution as detailed in #buildpacks/rfcs#002

### Additional context
- This feature should be documented as part of standard pack CLI commands documentation.
```

> After creating the issue, the maintainer should add the label `issues-created`  to the PR. Upon seeing that issue created, a member of the core team will merge it in.

Reviewer: As part of the merge process, should the core team member add references to the created issues in the `CNB Issue` field in the RFC meta section?

Maybe I am misunderstanding the original intention of this field (cc @hone)? In practice, we seem to leave it blank most of the time, but it could be powerful navigational tool.

Example: https://github.com/buildpacks/rfcs/blob/main/text/0044-pack-publish-buildpack.md#meta

Reviewer 2: yeah, that's the intention so we can reference the RFC discussion after it's been merged.

Author: Great point, @ekcasey. I'll add that in.

The current RFC process [defines](https://github.com/buildpacks/rfcs#rfc-process) that:
> the team member who merges the pull request should do the following ... Create a corresponding issue in the appropriate repo.

In practice, that adds extra work for the team member who merges in the PR, and it may slip through the cracks occasionally. Additionally, the sub-team maintainers are often the best people to create the issue, given their insight into how it should be implemented.

As an additional consideration, it is often hard to track work from the RFC proposal to the implementation, and difficult to ensure that it was completed. If we use [Github references](https://docs.github.com/en/github/writing-on-github/autolinked-references-and-urls#issues-and-pull-requests) to link the RFC PR to the issue on the appropriate repository, we can better track the work, and be notified when it is fully implemented.

# What it is
[what-it-is]: #what-it-is
This defines a process for sub-team maintainers to open issues on the relevant repositories, after FCP for the RFC concludes, before it is merged in.

# How it Works
[how-it-works]: #how-it-works

> When a PR has been accepted (after FCP has concluded), the sub-team maintainers make issues (tagging the RFC PR) in the relevant buildpack repos, and labels the PR `issues-created`. At that point, the RFC PR may be merged in, with references to the CNB issues added to the RFC [metadata](https://github.com/buildpacks/rfcs/blob/main/text/0044-pack-publish-buildpack.md#meta).

Reviewer 3: ```suggestion
When a PR has been accepted (after FCP has concluded), the sub-team maintainers make issues (tagging the RFC PR) in the relevant buildpack repos, and labels the PR `issues-created/<sub-team>` for their respective sub-team. Once all sub-team `issues-created/<sub-team>` labels are set, the RFC PR may be merged in, with references to the CNB issues added to the RFC [metadata](https://github.com/buildpacks/rfcs/blob/main/text/0044-pack-publish-buildpack.md#meta).
```

Author: Signed-off-by: David Freilich <dfreilich@vmware.com>
